pkg/cmd/pulumi/packagecmd/package_publish.go

```diff
@@ -0,0 +1,238 @@
+		return fmt.Errorf("failed to determine current project: %w", err)
+	}
+
+	b, err := login(ctx, project)
+	if err != nil {
+		return err
+	}
+
+	wd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+	sink := cmdutil.Diag()
+	pctx, err := plugin.NewContext(sink, sink, nil, nil, wd, nil, false, nil)
+	if err != nil {
+		return err
+	}
+	defer contract.IgnoreClose(pctx)
+
+	pkg, err := cmd.extractSchema(pctx, packageSrc, packageParams)
+	if err != nil {
+		return fmt.Errorf("failed to get schema: %w", err)
+	}
+
+	var publisher string
+	// If the publisher is set on the command line, use it.
+	if args.publisher != "" {
+		publisher = args.publisher
+	} else if pkg.Publisher != "" { // Otherwise, fall back to the publisher set in the package schema.
+		publisher = pkg.Publisher
+	} else { // As a last resort, try to determine the publisher from the default organization or fail if none is found.
+		publisher, err = cmd.defaultOrg(project)
+		if err != nil {
+			return fmt.Errorf("failed to determine default organization: %w", err)
+		}
+		if publisher == "" {
+			return errors.New("no publisher specified and no default organization found, please set a publisher in " +
+				"the package schema or set a default organization in your pulumi config")
+		}
+	}
+
+	name := pkg.Name
+	if name == "" {
+		return errors.New("no package name specified, please set one in the package schema")
+	}
+	var version semver.Version
+	if pkg.Version != nil {
+		version = *pkg.Version
+	} else {
+		return errors.New("no version specified, please set a version in the package schema")
+	}
+
+	json, err := pkg.MarshalJSON()
+	if err != nil {
+		return fmt.Errorf("failed to marshal schema: %w", err)
+	}
+
+	registry, err := b.GetPackageRegistry()
+	if err != nil {
+		return fmt.Errorf("failed to get package registry: %w", err)
+	}
+
+	// In the future we could slurp up the readme from the package source and add the necessary markdown headers.
+	readme, err := os.Open(args.readmePath)
+	if err != nil {
+		return fmt.Errorf("failed to open readme file: %w", err)
+	}
+	defer contract.IgnoreClose(readme)
+	readmeBytes := bytes.NewBuffer(nil)
+	if _, err := io.Copy(readmeBytes, readme); err != nil {
+		return fmt.Errorf("failed to read readme file: %w", err)
+	}
+
+	publishInput := backend.PackagePublishOp{
+		Source:    args.source,
+		Publisher: publisher,
+		Name:      name,
+		Version:   version,
+		Schema:    bytes.NewReader(json),
+		Readme:    readmeBytes,
+	}
+
+	if args.installDocsPath != "" {
+		installDocs, err := os.Open(args.installDocsPath)
+		if err != nil {
+			return fmt.Errorf("failed to open install docs file: %w", err)
+		}
+		defer contract.IgnoreClose(installDocs)
+		installDocsBytes := bytes.NewBuffer(nil)
+		if _, err := io.Copy(installDocsBytes, installDocs); err != nil {
+			return fmt.Errorf("failed to read install docs file: %w", err)
+		}
+		publishInput.InstallDocs = installDocsBytes
```

Comment:
Why are we doing this? `*os.File` is a `io.Reader`.
```suggestion
		if err != nil {
			return fmt.Errorf("failed to open install docs file: %w", err)
		}
		defer contract.IgnoreClose(installDocs)
		publishInput.InstallDocs = installDocs
```

Likewise for `readme`.

Reply:
We need to set the content-size header for s3 puts, for byte buffers the stdlib http client does that automatically.
For `os.File` it cannot do that because the size returned by `Stat()` is not reliable:
- Another process might modify the file by appending to it or truncating it
- If the file is a special file like a device file, pipe, or socket, its size may change dynamically
- On network filesystems, there could be caching or synchronization issues

By reading it into a buffer we circumvent these problems. I also do not think this has a meaningful impact on performance, the files should be rather small anyways (I'd be surprised if they're bigger than one MB).

Reply:
That's an excellent explanation. Can you put that into a comment in the file?

Reply:
Done!